addons/web/static/tests/views/form/form_view_tests.js

```diff
@@ -12573,4 +12573,91 @@ QUnit.module("Views", (hooks) => {
             "Unable to save"
         );
     });
+
+    QUnit.test(
+        "execute an action before and after each valid save in a form view",
+        async function (assert) {
+            const formView = registry.category("views").get("form");
+            class FormViewCustom extends formView.Controller {
+                async onSavedRecord(record) {
+                    assert.step(`onSavedRecord ${record.resId}`);
+                }
+
+                async onWillSaveRecord(record) {
+                    assert.step(`onWillSaveRecord ${record.resId}`);
+                }
+            }
+            registry.category("views").add(
+                "form",
+                {
+                    ...formView,
+                    Controller: FormViewCustom,
+                },
+                { force: true }
+            );
+
+            await makeView({
+                type: "form",
+                resModel: "partner",
+                resId: 1,
+                serverData,
+                arch: `<form><field name="foo" required="1"/></form>`,
+                mockRPC(route, args) {
+                    if (args.method === "write") {
+                        assert.step(`write ${args.args[0]}`);
+                    }
+                },
+            });
+
+            await editInput(target, "[name='foo'] input", "");
+            await clickSave(target);
+            assert.verifySteps(["onWillSaveRecord 1"]);
```

Comment:
this highlights a problem imo. This hook should only be called if the record will really be saved, don't you think?

Reply:
You wan't to do it after the checkValidity ?